evalml/pipelines/components/transformers/preprocessing/time_series_regularizer.py

```diff
@@ -0,0 +1,336 @@
+                    cleaned_y.iloc[misaligned_clean_ind] = y.iloc[
+                        misaligned_ind_original
+                    ]
+    return cleaned_x, cleaned_y
+
+
+class TimeSeriesRegularizer(Transformer):
+    """Transformer that regularizes an inconsistently spaced datetime column.
+
+    If X is passed in to fit/transform, the column `time_index` will be checked for an inferrable offset frequency. If
+    the `time_index` column is perfectly inferrable then this Transformer will do nothing and return the original X and y.
+
+    If X does not have a perfectly inferrable frequency but one can be estimated, then X and y will be reformatted based
+    on the estimated frequency for `time_index`. In the original X and y passed:
+     - Missing datetime values will be added and will have their corresponding columns in X and y set to None.
+     - Duplicate datetime values will be dropped.
+     - Extra datetime values will be dropped.
+     - If it can be determined that a duplicate or extra value is misaligned, then it will be repositioned to take the
+     place of a missing value.
+
+    This Transformer should be used before the `TimeSeriesImputer` in order to impute the missing values that were
+    added to X and y (if passed).
+
+    Args:
+        time_index (string): Name of the column containing the datetime information used to order the data.
+        random_seed (int): Seed for the random number generator. This transformer performs the same regardless of the random seed provided.
+    """
+
+    name = "Time Series Regularizer"
+    hyperparameter_ranges = {}
+    """{}"""
+
+    modifies_target = True
+    training_only = True
+
+    def __init__(
+        self, time_index=None, window_length=5, threshold=0.8, random_seed=0, **kwargs
+    ):
+        self.time_index = time_index
+        self.window_length = window_length
+        self.threshold = threshold
+        self.error_dict = {}
+        self.inferred_freq = None
+        self.debug_payload = None
+
+        parameters = {
+            "time_index": time_index,
+            "window_length": window_length,
+            "threshold": threshold,
+        }
+        parameters.update(kwargs)
+
+        super().__init__(parameters=parameters, random_seed=random_seed)
+
+    def fit(self, X, y=None):
+        """Fits the TimeSeriesRegularizer.
+
+        Args:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features].
+            y (pd.Series, optional): The target training data of length [n_samples].
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: if self.time_index is None
+            TypeError: if the `time_index` column is not of type Datetime
+            ValueError: if X and y have different lengths
+            ValueError: if `time_index` in X does not have an offset frequency that can be estimated
+        """
+        if self.time_index is None:
```

Comment:
Not a dumb question at all, it's for the same reason that the `TimeSeriesFeaturizer` and `ProphetRegressor` require it as well, a combination of being primarily a time series class and to match existing design.

There's also an implicit understanding that we don't support a datetime index for passed dataframes (yet) because it's a column name value you have to set.

Reply:
Oh I see what you mean, I don't mind removing a default value for `time_index` for this. I think we should do it in a couple of other places to stress the requirement of a `time_index` being passed in as well

Reply:
I'll also add an exception checking if a column was passed in that doesn't exist